src/include/execution/plans/window_plan.h

```diff
@@ -0,0 +1,135 @@
+ public:
+  /**
+   * Construct a new WindowAggregationPlanNode.
+   * @param output_schema The output format of this plan node
+   * @param child The child plan to aggregate data over
+   * @param partition_by The partition by clause of the window aggregation
+   * @param order_by The order by clause of the window aggregation
+   * @param aggregates The expressions that we are aggregating
+   * @param window_agg_types The types that we are aggregating
+   */
+  WindowAggregationPlanNode(SchemaRef output_schema, AbstractPlanNodeRef child, int window_agg_id,
+                            std::vector<AbstractExpressionRef> columns, std::vector<AbstractExpressionRef> partition_by,
+                            std::vector<std::pair<OrderByType, AbstractExpressionRef>> order_by,
+                            AbstractExpressionRef aggregate, WindowAggregationType window_agg_type)
+      : AbstractPlanNode(std::move(output_schema), {std::move(child)}),
+        window_agg_id_(window_agg_id),
+        columns_(std::move(columns)),
+        partition_by_(std::move(partition_by)),
+        order_by_(std::move(order_by)),
+        aggregate_(std::move(aggregate)),
+        window_agg_type_(window_agg_type) {}
+
+  /** @return The type of the plan node */
+  auto GetType() const -> PlanType override { return PlanType::Window; }
+
+  /** @return the child of this aggregation plan node */
+  auto GetChildPlan() const -> AbstractPlanNodeRef {
+    BUSTUB_ASSERT(GetChildren().size() == 1, "Window Aggregation expected to only have one child.");
+    return GetChildAt(0);
+  }
+
+  /** @return The idx'th partition by expression */
+  auto GetPartitionByAt(uint32_t idx) const -> const AbstractExpressionRef & { return partition_by_[idx]; }
+
+  /** @return The partition by expressions */
+  auto GetPartitionBys() const -> const std::vector<AbstractExpressionRef> & { return partition_by_; }
+
+  /** @return The idx'th order by expression */
+  auto GetOrderByAt(uint32_t idx) const -> const std::pair<OrderByType, AbstractExpressionRef> & {
+    return order_by_[idx];
+  }
+
+  /** @return The order by expressions */
+  auto GetOrderBys() const -> const std::vector<std::pair<OrderByType, AbstractExpressionRef>> & { return order_by_; }
+
+  /** @return The aggregate expression */
+  auto GetAggregate() const -> const AbstractExpressionRef & { return aggregate_; }
+
+  /** @return The window aggregate type */
+  auto GetWindowAggregateType() const -> const WindowAggregationType & { return window_agg_type_; }
+
+  static auto InferWindowSchema(const std::vector<AbstractExpressionRef> &columns, int window_agg_id) -> Schema;
+
+  BUSTUB_PLAN_NODE_CLONE_WITH_CHILDREN(WindowAggregationPlanNode);
+
+  /** The Window Aggregation column index */
+  int window_agg_id_;
+
+  /** all other columns expressions */
+  std::vector<AbstractExpressionRef> columns_;
+
+  /** The PARTITION BY expressions */
+  std::vector<AbstractExpressionRef> partition_by_;
```

Comment:
shouldn't this be a vector of "a vector of expressions"? if there are multiple aggregation functions with different partitions and order-by.

Reply:
Currently it only supports one window-function in select list. But I can extend it to support multiple ones, see below.

And I'm thinking about the design of executors, in normal aggregations it is fixed as an initial value first and a combine function to let students implementing the calculation of aggs based on a map. For window agg, are we gonna give similar design or we make it more flexible to define starter code that emits tuples and let students decide their implementation?

Reply:
We can simply given them three empty functions: constructor, init, and next. I’m also planning to strip all code from the current aggregation executor.